grype/matcher/apk/matcher.go

```diff
@@ -23,39 +26,115 @@ func (m *Matcher) Type() match.MatcherType {
 func (m *Matcher) Match(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
 	var matches = make([]match.Match, 0)
 
-	// map {  CVE string : []match }
-	var secDbCandidates = make(map[string][]match.Match)
-
 	// find Alpine SecDB matches for the given package name and version
 	secDbMatches, err := common.FindMatchesByPackageDistro(store, d, p, m.Type())
 	if err != nil {
 		return nil, err
 	}
 
-	for _, secDbMatch := range secDbMatches {
-		secDbCandidates[secDbMatch.Vulnerability.ID] = append(secDbCandidates[secDbMatch.Vulnerability.ID], secDbMatch)
+	cpeMatches, err := m.cpeMatchesWithoutSecDbFixes(store, d, p)
+	if err != nil {
+		return nil, err
 	}
 
-	// find NVD matches specific to the given package name and version
-	var cpeCandidates = make(map[string][]match.Match)
+	// keep all secdb matches, as this is an authoritative source
+	matches = append(matches, secDbMatches...)
+
+	// keep only unique CPE matches
+	matches = append(matches, deduplicateMatches(secDbMatches, cpeMatches)...)
+
+	return matches, nil
+}
+
+func (m *Matcher) cpeMatchesWithoutSecDbFixes(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	// find CPE-indexed vulnerability matches specific to the given package name and version
 	cpeMatches, err := common.FindMatchesByPackageCPE(store, p, m.Type())
 	if err != nil {
 		return nil, err
 	}
 
-	for _, cpeMatch := range cpeMatches {
-		cpeCandidates[cpeMatch.Vulnerability.ID] = append(cpeCandidates[cpeMatch.Vulnerability.ID], cpeMatch)
+	cpeMatchesByID := matchesByID(cpeMatches)
+
+	// remove cpe matches where there is an entry in the secDB for the particular package-vulnerability pairing, and the
+	// installed package version is >= the fixed in version for the secDB record.
+	secDbVulnerabilities, err := store.GetByDistro(*d, p)
```

Comment:
This returns all from the distro that match the package, but not doing an actual constraint check, correct?

Reply:
correct 👍